components/suggest/src/db.rs

```diff
@@ -123,14 +124,29 @@ impl<'a> SuggestDao<'a> {
 
     /// Fetches suggestions that match the given keyword from the database.
     pub fn fetch_by_keyword(&self, keyword: &str) -> Result<Vec<Suggestion>> {
+        let pocket_keyword = match keyword.contains(' ') {
+            true => keyword.to_string(),
+            false => keyword.to_owned() + " ",
```

Comment:
Oooh, I really like Nan's suggestion of using `Cow` that way!

But I think that this logic won't handle single-word low-confidence keywords correctly. For example, I don't think that `workaholism` will match the Pocket suggestion in the test now—but it should. Is that right?

IIUC, another way to write our algorithm for matching low-confidence keywords is:

1. The head (first word) of the query must be the same as the low-confidence keyword's head.
2. The tail (anything after the first word) of the query—if any—should be a prefix of the keyword's tail.

So all these queries should match the Pocket suggestion: `workaholism`, `soft` (first word of `soft life`), `soft l`, `soft li` (first words match; `l` and `li` are prefixes of `life`).

And these shouldn't, because their first words don't match: `work`, `sof`, etc.

Reply:
One way we could make this work is to store the head and tail in separate columns when we insert them into the table. Then, when we fetch suggestions, we could split the user's query the same way, and match the head and tail prefix.

Here's a quick sketch of how that all could work! Let's say we have two suggestions, and all their low-confidence keywords start with `breakfast`:

```text
title: "How to Build a Better Breakfast"
keywords: ["breakfast"]

title: Your New Favorite Breakfast Sandwich
keywords: ["breakfast crunch wrap", "breakfast torta"]
```

If we stored those suggestions and keywords in a schema like this:

```sql
CREATE TABLE suggestions(id INTEGER PRIMARY KEY, title TEXT NOT NULL);
CREATE TABLE keywords(keyword_head TEXT NOT NULL, keyword_tail TEXT NOT NULL, suggestion_id INTEGER NOT NULL REFERENCES suggestions(id), PRIMARY KEY (keyword_head, keyword_tail));

INSERT INTO suggestions(id, title) VALUES(1, 'How to Build a Better Breakfast');
INSERT INTO suggestions(id, title) VALUES(2, 'Your New Favorite Breakfast Sandwich');

INSERT INTO keywords(keyword_head, keyword_tail, suggestion_id) VALUES('breakfast', '', 1);
INSERT INTO keywords(keyword_head, keyword_tail, suggestion_id) VALUES('breakfast', 'crunch wrap', 2);
INSERT INTO keywords(keyword_head, keyword_tail, suggestion_id) VALUES('breakfast', 'torta', 2);
```

Here's how a query plan for `breakfast` would look:

```text
sqlite> SELECT s.id, s.title FROM suggestions s JOIN keywords k ON k.suggestion_id = s.id WHERE k.keyword_head = 'breakfast' AND k.keyword_tail BETWEEN '' AND '' || x'ffff' GROUP BY s.id, s.title;
QUERY PLAN
|--SEARCH k USING INDEX sqlite_autoindex_keywords_1 (keyword_head=? AND keyword_tail>? AND keyword_tail<?)
|--SEARCH s USING INTEGER PRIMARY KEY (rowid=?)
`--USE TEMP B-TREE FOR GROUP BY
1|How to Build a Better Breakfast
2|Your New Favorite Breakfast Sandwich
```

(To play around with that schema in the `sqlite3` shell, I saved the schema into a file, then ran `sqlite3 -init file.sql` to load it).

Since `Your New Favorite Breakfast Sandwich` has two keywords that start with `breakfast`, we need a `GROUP BY` or a `DISTINCT` to filter out duplicates—or we'll get two rows for it, one for each of its keywords. But the search over `keywords` can use all columns of the primary key (that's the `sqlite_autoindex_keywords_1` automatic index), so the most expensive part of the query is still pretty efficient. By the time we build the temporary b-tree for the `GROUP BY`, we should have reduced the number of rows to a small handful.

The query plan for `breakfast cru` looks the same, and finds just the one suggestion:

```text
sqlite> SELECT s.id, s.title FROM suggestions s JOIN keywords k ON k.suggestion_id = s.id WHERE k.keyword_head = 'breakfast' AND k.keyword_tail BETWEEN 'cru' AND 'cru' || x'ffff' GROUP BY s.id, s.title;
QUERY PLAN
|--SEARCH k USING INDEX sqlite_autoindex_keywords_1 (keyword_head=? AND keyword_tail>? AND keyword_tail<?)
|--SEARCH s USING INTEGER PRIMARY KEY (rowid=?)
`--USE TEMP B-TREE FOR GROUP BY
2|Your New Favorite Breakfast Sandwich
```

And so does `breakfast torta`:

```text
sqlite> SELECT s.id, s.title FROM suggestions s JOIN keywords k ON k.suggestion_id = s.id WHERE k.keyword_head = 'breakfast' AND k.keyword_tail BETWEEN 'torta' AND 'torta' || x'ffff' GROUP BY s.id, s.title;
QUERY PLAN
|--SEARCH k USING INDEX sqlite_autoindex_keywords_1 (keyword_head=? AND keyword_tail>? AND keyword_tail<?)
|--SEARCH s USING INTEGER PRIMARY KEY (rowid=?)
`--USE TEMP B-TREE FOR GROUP BY
2|Your New Favorite Breakfast Sandwich
```

Reply:
Wow, that's some really cool SQL magic there! 👏

I'd like to offer another approach: How about we move this qualifying logic over to the row handler closure?

We currently only fetch data from other tables and then assemble the suggestion accordingly in that closure. I think we can also use that as the __post-processor__ for each candidate. Specifically, the post-processor can transform, filter, and augment each row if needed. Two benefits I can see here:
- In certain cases, it's much easier to do stuff in native code than in SQL
- To keep those keywords table schema lean and generic. Imagine if we need to change the matching logic for the low-confidence keywords, we can still do prefix queries in SQL and apply more specific matching logic in the post-processor. No change is needed to the schema!

WDYT?

Reply:
> How about we move this qualifying logic over to the row handler closure?

Filtering in the row handler is definitely a great option, with a small result set. For example, if our query returns 5 rows, and we filter all of them out—that's OK, because we only had to scan and read 5 rows. But if our query returns hundreds or thousands of rows, filtering in the row handler isn't as efficient—we still end up scanning and reading in all those rows, and then filtering them out in code.

For low-confidence Pocket suggestions, this is a bit tricky to do naively, because we want the whole first word to match, and only match prefixes for subsequent words. If we wanted to keep the schema super generic, with a single `keyword` column, I think we'd need to do a prefix match unconditionally. So, if the user searches for `b` or `br`, we'd return suggestions for all keywords that start with `b` or `br`—and then filter out all of them, because `b` and `br` aren't words.

We could combine the approaches, through—a less generic schema, but also doing more work in Rust code instead of SQL:

1. Split each keyword into `keyword_head` and `keyword_tail` (I'm realizing `keyword_prefix` and `keyword_suffix` might actually be better names! 😅), like I suggested, and...
2. Always filter on `keyword_prefix` in the SQL query, then filter on `keyword_suffix` in the row handler, like you suggested!

I think this would also line up nicely with @0c0w3's suggestion to use a single table to store the keywords, and have a column for whether it's low or high-confidence.

So we'd have a schema like:

```sql
CREATE TABLE pocket_keywords(
    keyword_prefix TEXT NOT NULL,
    keyword_suffix TEXT NOT NULL DEFAULT "",
    confidence_type INTEGER NOT NULL,
    suggestion_id INTEGER NOT NULL REFERENCES suggestions(id),
    PRIMARY KEY (keyword_prefix, keyword_suffix)
);
```

With the data from [Drew's example ](https://github.com/mozilla/application-services/pull/5841#issuecomment-1736521129)like:

```
keyword_prefix | keyword_suffix | confidence_type
---------------+----------------+----------------
esther         | the wonder pig | LOW
wildlife       | stories        | HIGH
```

(Assume `LOW` and `HIGH` are integer constants).

And then, in the row handler, we'd have logic like:

```rs
let suffixes_match = match row.get("confidence_type")? {
    LOW => row.get("keyword_suffix")?.starts_with(user_keyword_suffix),
    HIGH => row.get("keyword_suffix")? == user_keyword_suffix,
};
```

Let's say the user typed in `esther th`. Splitting that on the first space, the prefix is `esther`, and the suffix is `th`. The query for that would look like this:

```sql
SELECT s.*, k.keyword_suffix, k.confidence_type
FROM suggestions s
JOIN pocket_keywords k ON k.suggestion_id = s.id
WHERE k.keyword_prefix = 'esther'
```

...Then, our row handler would see that it's a low-confidence keyword, and take the `LOW` branch. `the wonder pig` starts with `th`, so it's a match.

Now let's say they typed in `wildlife stories`. The prefix is `wildlife`; the suffix is `stories`. The query for that looks the same as for `esther th`:

```sql
SELECT s.*, k.keyword_suffix, k.confidence_type
FROM suggestions s
JOIN pocket_keywords k ON k.suggestion_id = s.id
WHERE k.keyword_prefix = 'wildlife'
```

...Then our row handler sees that it's a high-confidence keyword, so it takes the `HIGH` branch. `stories` == `stories`, so it's a match.

Reply:
Nice, that'll work like a charm!